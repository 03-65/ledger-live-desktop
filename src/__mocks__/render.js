import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import createHistory from 'history/createHashHistory'

import createStore from 'renderer/createStore'

import theme from 'styles/theme'

export default function render(component, state) {
  const history = createHistory()
  const store = createStore(history, state)
  return renderer
    .create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </Provider>,
    )
    .toJSON()
}
