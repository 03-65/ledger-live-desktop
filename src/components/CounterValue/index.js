// @flow

import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import type { CryptoCurrency } from '@ledgerhq/live-common/lib/types'

import { counterValueCurrencySelector } from 'reducers/settings'
import { calculateCounterValueSelector } from 'reducers/counterValues'

import FormattedVal from 'components/base/FormattedVal'

type Props = {
  // wich market to query
  ticker: string,

  // when? if not given: take latest
  date?: Date,

  // from reducers
  counterValueCurrency: CryptoCurrency,
  value: number,
}

const mapStateToProps = (state, props) => {
  const { ticker, value, date } = props

  // TODO: in wallet-common, stop using currency.
  // always use ticker and remove that hack
  let { currency } = props
  if (!currency && ticker) {
    currency = generateFakeCurrency(ticker)
  } else if (currency) {
    console.warn('`currency` is deprecated in CounterValue. use `ticker` instead.') // eslint-disable-line no-console
  }

  const counterValueCurrency = counterValueCurrencySelector(state)
  const counterValue =
    !counterValueCurrency || !currency
      ? 0
      : calculateCounterValueSelector(state)(currency, counterValueCurrency)(value, date)

  return {
    counterValueCurrency,
    value: counterValue,
  }
}

class CounterValue extends PureComponent<Props> {
  static defaultProps = {
    value: 0,
    date: undefined,
  }

  render() {
    const { value, counterValueCurrency, date, ...props } = this.props
    return (
      <FormattedVal
        val={value}
        fiat={counterValueCurrency.units[0].code}
        showCode
        alwaysShowSign
        {...props}
      />
    )
  }
}

function generateFakeCurrency(ticker) {
  return {
    ticker,
    units: [
      {
        code: ticker,

        // unused
        name: 'fake-unit',
        magnitude: 0,
      },
    ],

    // unused
    coinType: 0,
    color: '#000',
    name: 'fake-coin',
    scheme: 'bitcoin',
  }
}

export default connect(mapStateToProps)(CounterValue)
