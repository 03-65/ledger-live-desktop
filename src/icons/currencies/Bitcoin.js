import React from 'react'

export default props => (
  <svg viewBox="0 0 16 16" {...props}>
    <path
      fill="currentColor"
      d="M7.61970553 0h1.70000005v3.05846154H7.61970553V0zm1.70000005 0v3.05846154H7.61970553V0h1.70000005zM7.81303887 12.9415385h1.70000004V16H7.81303887v-3.0584615zm1.70000004 0V16H7.81303887v-3.0584615h1.70000004zM4.4063722 0h1.70000005v3.05846154H4.4063722V0zm1.70000005 0v3.05846154H4.4063722V0h1.70000005zM4.59970553 12.9415385h1.70000005V16H4.59970553v-3.0584615zm1.70000005 0V16H4.59970553v-3.0584615h1.70000005zM2.6663722 8.81307695V1.94999998h.85000002l6.13827333.00156806c1.80778725.10996034 3.21199965 1.5663186 3.11839335 3.25458581l-.0017039.27851419c.110634 1.74464117-1.2981636 3.21696546-3.16829611 3.32840891H2.6663722zm.85000002 0l.85000003-.85000003v4.31926938l5.60765891.000255c.95057664-.0249932 1.67559124-.7265334 1.65900774-1.5564475l.0002249-.3703244c.0187355-.81415185-.7068616-1.51805139-1.63689158-1.54275245h-6.48zm7.55666668-3.27461541l.0012684-.37872715c.042116-.77004688-.6196246-1.45636138-1.47126841-1.50973437H4.36637225v3.46311513l5.20484808.0002342c.89861507-.06354201 1.55224847-.7529014 1.50181857-1.57488781zM4.36637225 7.11311515l5.18540306.00150901c.00432951-.00026159.15996192-.00067981.46689729-.00125468 1.859875.04881191 3.3551536 1.49937703 3.3143663 3.26201512l-.0001732.3336104c.0359164 1.7788337-1.4578252 3.2242112-3.33649348 3.2733127H2.6663722V7.1130769h.85000002l.85000003.00003825zm0 .00017966v.84978211l-.85000003-.85000002.85000003.00021791z"
    />
  </svg>
)
