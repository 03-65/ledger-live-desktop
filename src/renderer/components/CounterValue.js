// @flow

import type { BigNumber } from "bignumber.js";
import React from "react";
import { connect } from "react-redux";
import type { Currency } from "@ledgerhq/live-common/lib/types";
import {
  counterValueCurrencySelector,
  exchangeSettingsForPairSelector,
  intermediaryCurrency,
} from "~/renderer/reducers/settings";
import CounterValues from "~/renderer/countervalues";
import FormattedVal from "~/renderer/components/FormattedVal";
import type { State } from "~/renderer/reducers";

type OwnProps = {
  // wich market to query
  currency: Currency,

  // when? if not given: take latest
  date?: Date,

  value: BigNumber,

  alwaysShowSign?: boolean,

  subMagnitude?: number,

  placeholder?: React$Node,

  prefix?: React$Node,
  suffix?: React$Node,
};

type StateProps = {|
  counterValueCurrency: Currency,
  value: ?number,
|};

type Props = {
  ...OwnProps,
  ...StateProps,
};

const mapStateToProps = (state: State, props: OwnProps) => {
  const { currency, value, date, subMagnitude } = props;
  const counterValueCurrency = counterValueCurrencySelector(state);
  const intermediary = intermediaryCurrency(currency, counterValueCurrency);
  const fromExchange = exchangeSettingsForPairSelector(state, { from: currency, to: intermediary });
  const toExchange = exchangeSettingsForPairSelector(state, {
    from: intermediary,
    to: counterValueCurrency,
  });
  const counterValue = CounterValues.calculateWithIntermediarySelector(state, {
    from: currency,
    fromExchange,
    intermediary,
    toExchange,
    to: counterValueCurrency,
    value,
    date,
    disableRounding: !!subMagnitude,
  });

  return {
    counterValueCurrency,
    value: counterValue,
  };
};

function CounterValue({
  value,
  counterValueCurrency,
  date,
  alwaysShowSign,
  placeholder,
  prefix,
  suffix,
  ...props
}: Props) {
  if (!value) {
    return placeholder || null;
  }

  console.log("!!!", props);
  return (
    <>
      {prefix || null}
      <FormattedVal
        {...props}
        val={value}
        unit={counterValueCurrency.units[0]}
        showCode
        alwaysShowSign={alwaysShowSign || true}
      />
      {suffix || null}
    </>
  );
}

export default connect(mapStateToProps)(CounterValue);
