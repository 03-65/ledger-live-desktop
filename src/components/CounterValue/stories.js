// @flow

import React from 'react'
import { getDefaultUnitByCoinType } from '@ledgerhq/currencies'
import { storiesOf } from '@storybook/react'
import { boolean, text } from '@storybook/addon-knobs'

import { CounterValue } from 'components/CounterValue'

const stories = storiesOf('Components', module)

const unit = getDefaultUnitByCoinType(0)

const counterValue = 'USD'
const counterValues = {
  'BTC-USD': {
    byDate: {
      '2018-01-09': 10000,
    },
    list: [['2018-01-09', 10000]],
  },
}

stories.add('CounterValue', () => (
  <CounterValue
    counterValue={counterValue}
    counterValues={counterValues}
    unit={unit}
    formatValue={boolean('formatValue', true)}
    value={Number(text('value', '100000000'))}
  />
))
