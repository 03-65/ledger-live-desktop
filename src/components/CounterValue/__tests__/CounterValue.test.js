// @flow

import React from 'react'
import { getDefaultUnitByCoinType } from '@ledgerhq/currencies'

import render from '__mocks__/render'
import CounterValue from '..'

describe('components', () => {
  describe('CounterValue', () => {
    it('basic', () => {
      const state = { counterValues: { BTC: { USD: { latest: 10e2 } } } }
      const unit = getDefaultUnitByCoinType(1)
      const component = <CounterValue ticker="BTC" unit={unit} value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('specifying ticker different from default', () => {
      const state = { counterValues: { LOL: { USD: { latest: 5e2 } } } }
      const unit = getDefaultUnitByCoinType(1)
      const component = <CounterValue ticker="LOL" unit={unit} value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('using countervalue different from default', () => {
      const state = {
        counterValues: { BTC: { EUR: { latest: 42 } } },
        settings: {
          counterValue: 'EUR',
        },
      }
      const unit = getDefaultUnitByCoinType(1)
      const component = <CounterValue ticker="BTC" unit={unit} value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('without countervalues populated', () => {
      const state = { counterValues: {} }
      const unit = getDefaultUnitByCoinType(1)
      const component = <CounterValue ticker="BTC" unit={unit} value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('with time travel whith date in countervalues', () => {
      const state = { counterValues: { BTC: { USD: { '2018-01-01': 20e2 } } } }
      const unit = getDefaultUnitByCoinType(1)

      const date = new Date('2018-01-01')
      const component = <CounterValue ticker="BTC" unit={unit} value={1} date={date} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('with time travel whith date not in countervalues', () => {
      const state = { counterValues: { BTC: { USD: { '2018-01-01': 20e2 } } } }
      const unit = getDefaultUnitByCoinType(1)

      const date = new Date('2018-01-02')
      const component = <CounterValue ticker="BTC" unit={unit} value={1} date={date} />
      const tree = render(component, state)

      // TODO: actually it returns 0 when date is not in countervalues
      // do we want to use closest past value instead?
      expect(tree).toMatchSnapshot()
    })
  })
})
