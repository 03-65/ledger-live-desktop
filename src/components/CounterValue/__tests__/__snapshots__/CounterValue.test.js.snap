// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CounterValue basic 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 bJwMKx"
  color="#66be54"
>
  + USD 10.00
</div>
`;

exports[`components CounterValue specifying ticker different from default 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 bJwMKx"
  color="#66be54"
>
  + USD 5.00
</div>
`;

exports[`components CounterValue using countervalue different from default 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 huzgmt"
  color={undefined}
>
  + EUR 0.42
</div>
`;

exports[`components CounterValue with time travel whith date in countervalues 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 bJwMKx"
  color="#66be54"
>
  + USD 20.00
</div>
`;

exports[`components CounterValue with time travel whith date not in countervalues 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 bJwMKx"
  color="#66be54"
>
  + USD 0.00
</div>
`;

exports[`components CounterValue without countervalues populated 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 bJwMKx"
  color="#66be54"
>
  + USD 0.00
</div>
`;
